mempool/src/pool/feerate_points.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright (c) 2021-2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use std::{
+    cmp::{Eq, Ord},
+    collections::BTreeMap,
+    ops::{Add, Div, Mul, Sub},
+};
+
+use common::primitives::Amount;
+
+use super::fee::Fee;
+
+pub fn linear_interpolation<X, Y>(p0: (X, Y), p1: (X, Y), x: X) -> Option<Y>
+where
+    X: Add<Output = X> + Sub<Output = X> + Eq + Ord + Copy,
+    Y: Mul<Output = Y> + Add<Output = Y> + Div<Output = Y> + Copy + From<X>,
+{
+    if p0.0 == p1.0 {
+        // Avoid division by zero
+        return if p0.0 == x { Some(p0.1) } else { None };
+    }
+
+    if x < p0.0 || x > p1.0 {
+        // The interpolation factor is outside the range
+        None
+    } else {
+        let scaled_v1 = p0.1 * Y::from(p1.0 - x);
+        let scaled_v2 = p1.1 * Y::from(x - p0.0);
+        let total_scale = Y::from(p1.0 - p0.0);
+
+        let interpolated_value = (scaled_v1 + scaled_v2) / total_scale;
+        Some(interpolated_value)
+    }
+}
+
+pub fn find_interpolated_value<T: From<Fee> + Into<Fee> + Copy>(
+    map: &BTreeMap<usize, T>,
+    key: usize,
+) -> Option<T> {
+    match map.get(&key) {
+        Some(value) => Some(*value),
+        None => {
+            let (k1, left_value) = map.range(..key).next_back()?;
+            let (k2, right_value) = map.range(key..).next()?;
+            let left_value: Fee = (*left_value).into();
+            let right_value: Fee = (*right_value).into();
+
+            let interpolated_value = linear_interpolation(
+                (*k1 as u64, left_value.into_atoms()),
+                (*k2 as u64, right_value.into_atoms()),
+                key as u64,
+            )?;
+
+            Some(Fee::new(Amount::from_atoms(interpolated_value)).into())
+        }
+    }
+}
+
+// Generate equidistant points to evaluate the mathematical function at for interpolation
+pub fn generate_equidistant_span(first: usize, last: usize, n: usize) -> Vec<usize> {
+    if first == last || n < 2 {
+        return vec![first];
+    }
+
+    let mut points = Vec::with_capacity(n);
+    points.push(first);
+
+    let step = (last - first) / (n - 1);
+
+    for i in 1..(n - 1) {
+        points.push(first + step * i);
+    }
```

Comment:
I think the rounding error in `step` calculation may be quite significant here in certain cases. Let's say last = 27, first = 10, n = 10. Then step = floor(17 / 9) = 1, so we will generate [10, 11, 12, 13, 14, 15, 16, 17, 18, 27]. There's a significant jump between the second last and the last element. Even more extreme case, if last = 18 (other inputs unchanged), we get step = floor(8 / 9) = 0, generating [10, 10, 10, 10, 10, 10, 10, 10, 10, 17].

Evaluating the whole thing at each step (`i * (last - first) / (n - 1)`) should avoid the issue but then you may need to deal with overflow or use a wider type.

Not sure how much this is an issue in practice for calculating feerate points but the function looks quite generic and if somebody uses it in another setting, it may be a problem.